Source/WebCore/platform/audio/SharedAudioDestination.cpp

```diff
@@ -0,0 +1,249 @@
+    static Ref<SharedAudioDestinationAdapter> ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction);
+
+    void addRenderer(SharedAudioDestination&, CompletionHandler<void(bool)>&&);
+    void removeRenderer(SharedAudioDestination&, CompletionHandler<void(bool)>&&);
+
+    unsigned framesPerBuffer() const
+    {
+        return m_workBus->length();
+    }
+
+private:
+    using AdapterKey = std::tuple<unsigned, float>;
+    using AdapterMap = HashMap<AdapterKey, WeakPtr<SharedAudioDestinationAdapter>>;
+    static AdapterMap& sharedMap();
+
+    SharedAudioDestinationAdapter(unsigned numberOfOutputChannels, AudioDestinationCreationFunction&&);
+
+    void render(AudioBus* sourceBus, AudioBus* destinationBus, size_t framesToProcess, const AudioIOPosition& outputPosition) final;
+    void isPlayingDidChange() final { }
+
+    void configureRenderThread(CompletionHandler<void(bool)>&&);
+
+    Ref<AudioDestination> m_destination;
+    Ref<AudioBus> m_workBus;
+    Ref<WorkQueue> m_configurationQueue;
+
+    Lock m_renderLock;
+    Semaphore m_configurationSemaphore;
+
+    // Only access on MainThread
+    Vector<RefPtr<SharedAudioDestination>> m_renderers;
+    bool m_started { false };
+
+    // Protected by m_renderLock:
+    bool m_needsConfiguration { true };
+    Vector<RefPtr<SharedAudioDestination>> m_newRenderers;
+    Vector<RefPtr<SharedAudioDestination>> m_configuredRenderers;
+};
+
+auto SharedAudioDestinationAdapter::sharedMap() -> AdapterMap&
+{
+    static NeverDestroyed<AdapterMap> map;
+    return map;
+}
+
+Ref<SharedAudioDestinationAdapter> SharedAudioDestinationAdapter::ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction)
+{
+    auto key = std::make_tuple(numberOfOutputChannels, sampleRate);
+    auto results = sharedMap().find(key);
+    if (results != sharedMap().end() && results->value)
+        return Ref { *results->value };
+
+    Ref newAdapter = adoptRef(*new SharedAudioDestinationAdapter(numberOfOutputChannels,  WTFMove(ensureFunction)));
+    sharedMap().set(key, newAdapter);
+    return newAdapter;
+}
+
+SharedAudioDestinationAdapter::SharedAudioDestinationAdapter(unsigned numberOfOutputChannels, AudioDestinationCreationFunction&& ensureFunction)
+    : m_destination { ensureFunction(*this) }
+    , m_workBus { AudioBus::create(numberOfOutputChannels, AudioUtilities::renderQuantumSize).releaseNonNull() }
+    , m_configurationQueue { WorkQueue::create("SharedAudioDestinationAdapter configuration queue") }
+    , m_configurationSemaphore(0)
+{
+}
+
+void SharedAudioDestinationAdapter::addRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        ASSERT(!m_renderers.contains(renderer));
```

Comment:
How is this safe? what guarantees that `this` is still alive when the lambda runs?

Reply:
Ah looks like I dropped that part during a refactor. There should be a sync dispatch to the configuration queue in the destructor of `SharedAudioDestinationAdapter`.

Reply:
This would work but it seems fragile? Why can't we protect in lambdas?

Reply:
We have well established smart pointer adoption patterns. It would be good to follow them. And if we can't it would be good to understand why so we can tweak our patterns accordingly.

Reply:
Okay, I'll add a WeakPtr/RefPtr lambda check.